### PR TITLE
Expose Laravel Loop Object To Column

### PR DESCRIPTION
**Pull request by Nuhel**
Sometime we need to show the Serial Number/Iteration Number in table. I could not find any way to do this with filament table. (May be there is a way, if so i will close this PR)

This PR Expose Laravel Loop Object To Columns,

We can use it like below

```php

Tables\Columns\TextColumn::make('sl')->getStateUsing(function ($loop){
      return $loop->iteration;
}),
```

Also as it is laravel's foreach loop object, We can access every other property like
index, remaining, count etc.

### CODE REVIEW

**Comment by danharrin** [comment]
Instead of `loop`, please use `rowIndex` and start from 0

**Comment by Nuhel** [comment]
> Instead of `loop`, please use `rowIndex` and start from 0

You mean instead of loop object just pass an int that starts from 0?

**Comment by danharrin** [comment]
well at the moment you're passing in the column loop object and not the row loop object? i think? which is wrong right?

**Comment by Nuhel** [comment]
> well at the moment you're passing in the column loop object and not the row loop object? i think? which is wrong right?

Yes, if we pass the loop object, we will have some extra options like remaining, count, depth, last etc, which will help if we need those.

But it's your call , i will update the PR if you think loop object is not a good fit.

**Comment by danharrin** [comment]
I agree now that the loop object is fine, but please test your code. I don't think it works, as you're passing in the column loop and not the row loop. For clarity we can name it `rowLoop()` as well, but I think it still needs fixing

**Comment by Nuhel** [comment]
> I agree now that the loop object is fine, but please test your code. I don't think it works, as you're passing in the column loop and not the row loop. For clarity we can name it `rowLoop()` as well, but I think it still needs fixing

Yes I was passing column loop -_- .

Test With normal Table Layout.
Screenshot 2022-11-07 at 7 19 55 PM

Test with Split and Stack
Screenshot 2022-11-07 at 7 19 26 PM

**Comment by danharrin** [comment]
I've just refactored a bit, does it still work?

**Comment by Nuhel** [comment]
> I've just refactored a bit, does it still work?

Not Working With Stack & Split.
Filament\Tables\Columns\Column::rowLoop(): Argument #1 ($loop) must be of type ?stdClass, Closure given,

**Comment by Nuhel** [comment]
> I've just refactored a bit, does it still work?

{{ $layoutComponent->viewData(['recordKey' => $recordKey, 'rowLoop' => $rowLoop]) }}

rowLoop has to be passed in viewData.

**Comment by danharrin** [comment]
@Nuhel how about passing it into the object instead and then using `$getRowLoop()` in the view? I've added the trait to the Component object

**Comment by Nuhel** [comment]
> @Nuhel how about passing it into the object instead and then using `$getRowLoop()` in the view? I've added the trait to the Component object

May be i did not get you, can you elaborate please?

**Comment by danharrin** [comment]
I've made the changes, test again please

**Comment by Nuhel** [comment]
> I've made the changes, test again please

Yes working. 

Should I update documentation? And I am thinking to add below function to CanFormatState Trait

```php
public function rowIndex(bool $fromZero = false): static
{
    $this->getStateUsing(static function ($rowLoop) use($fromZero): ?string {
        return $rowLoop->{ $fromZero? "index":"iteration"};
    });

    return $this;
}
```

File: packages/tables/src/Columns/Concerns/CanFormatState.php [hunk 1]
```diff
@@ -194,4 +195,13 @@ public function getTimezone(): string
     {
         return $this->evaluate($this->timezone) ?? config('app.timezone');
     }
+
+    public function rowIndex(bool $fromZero = false): static
```
Comment (danharrin) [inline packages/tables/src/Columns/Concerns/CanFormatState.php]: Please move this method to the TextInput class

**Pull request merged**